File: DiveIntoDesignPatterns/creational/AbstractFactory/sourcemaking.py
#py 3.8


class Widget(object):

    def draw(self):
        raise NotImplementedError


class LinuxButton(Widget):

    def draw(self):
        print("LinuxButton")


class LinuxMenu(Widget):

    def draw(self):
        print("LinuxMenu")


class WindowsButton(Widget):

    def draw(self):
        print("WindowsButton")


class WindowsMenu(Widget):

    def draw(self):
        print("WindowsMenu")


class Factory(object):

    def create_button(self):
        raise NotImplementedError

    def create_menu(self):
        raise NotImplementedError


class LinuxFactory(Factory):

    def create_button(self):
        return LinuxButton()

    def create_menu(self):
        return LinuxMenu()


class WindowsFactory(Factory):

    def create_button(self):
        return WindowsButton()

    def create_menu(self):
        return WindowsMenu()


class Client(object):

    def __init__(self, factory: Factory):
        self.factory = factory

    def display_window_one(self):
        factories = [self.factory.create_button(),
                     self.factory.create_menu()]
        factories[0].draw()
        factories[1].draw()

    def display_window_two(self):
        factories = [self.factory.create_menu(),
                     self.factory.create_button()]
        factories[0].draw()
        factories[1].draw()

    def draw(self):
        factory = self.factory.create_button()
        factory.draw()

        self.display_window_one()
        self.display_window_two()


def main(platform):
    if platform == "linux":
        factory = LinuxFactory()
    else:
        factory = WindowsFactory()

    client = Client(factory)
    client.draw()


if __name__ == '__main__':
    main("linux")

File: DiveIntoDesignPatterns/creational/AbstractFactory/sourcemaking.cpp
#include <iostream>
#define LINUX


// reference: https://sourcemaking.com/design_patterns/abstract_factory/cpp/before-after
using namespace std;


class Widget {
public:
    virtual void draw() = 0;
};


class LinuxButton : public Widget {
public:
    void draw() { cout << "LinuxButton\n"; }
};


class LinuxMenu : public Widget {
public:
    void draw() { cout << "LinuxMenu\n"; }
};


class WindowsButton : public Widget {
public:
    void draw() { cout << "WindowsButton\n"; }
};


class WindowsMenu : public Widget {
public:
    void draw() { cout << "WindowsMenu\n"; }
};


/******************************************************************************
* Abstract factory defines methods to create all related products.            *
*******************************************************************************/
class Factory {
public:
    virtual Widget* create_button() = 0;
    virtual Widget* create_menu() = 0;
};


/******************************************************************************
* Each concrete factory corresponds to one product
* family. It creates all possible products of one kind.
*******************************************************************************/
class LinuxFactory : public Factory {
public:
    Widget* create_button() {
        return new LinuxButton;
    }
    Widget* create_menu() {
        return new LinuxMenu;
    }
};


class WindowsFactory : public Factory {
public:
    Widget* create_button() {
        return new WindowsButton;
    }
    Widget* create_menu() {
        return new WindowsMenu;
    }
};


/**
 * Client receives a factory object from its creator.
 *
 * All clients work with factories through abstract
 * interface. They don't know concrete classes of
 * factories. Because of this, you can interchange
 * concrete factories without breaking clients.
 *
 * Clients don't know the concrete classes of created
 * products either, since abstract factory methods
 * returns abstract products.
 */
class Client {
private:
    Factory* factory;

public:
    Client(Factory* f) {
        factory = f;
    }

    void draw() {
        Widget* w = factory->create_button();
        w->draw();
        display_window_one();
        display_window_two();
    }

    void display_window_one() {
        Widget* w[] = {
            factory->create_button(),
            factory->create_menu()
        };
        w[0]->draw();
        w[1]->draw();
    }

    void display_window_two() {
        Widget* w[] = {
            factory->create_menu(),
            factory->create_button()
        };
        w[0]->draw();
        w[1]->draw();
    }
};


/**
 * Now the nasty switch statement is needed only once to
 * pick and create a proper factory. Usually that's
 * happening somewhere in program initialization code.
 */
int test_abstract_factory_widget() {
    Factory* factory;
#ifdef LINUX
    factory = new LinuxFactory;
#else // WINDOWS
    factory = new WindowsFactory;
#endif

    Client* c = new Client(factory);
    c->draw();

    return 0;
}